src/client/shader.h

```diff
@@ -239,11 +243,13 @@ class IShaderSource {
 	 * @param name name of the shader (directory on disk)
 	 * @param input_const primary key constants for this shader
 	 * @param base_mat base material to use
+	 * @param uniforms_cb_override override of the Uniform Setter callback
```

Comment:
why do this? from a first look this seems to bypass the abstraction and would be bad design

Reply:
The goal is to avoid to set all the Uniforms that exist in the program for shaders that don't need it.
I find it cleaner that the _shadow shader_ only has its related uniforms set, i.e. only one _ShadowDepthShaderCB_ is called on it. And in the other direction too: node_objects don't need _ShadowDepthShaderCB_ to set uniform such as `LightMVP`.

Most of the shaders have common UniformSetter callbacks: node_shader and object_shader .

I am pretty sure this was the issue discussed in `@Liso's disclaimer` comment.


I would even argue that the class SourceShader should not even owns all callbacks and factories given to shaders by default, by each call to `getShader` should pass which UniformSetters should be called.

Reply:
I think we can agree that the shadow shaders are different from all other shaders.
What would the minimal change to have them use only respective uniforms? And then we can discuss a general solution - if needed - in another PR.

Reply:
To me the current state of the PR is the minimal change to have _only respective uniforms_.  Another minimal change could be to have _all uniforms for all shaders_ (what I implemented in the first commits of this PR).  

Open to any other alternative or suggestion.